Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs

```diff
@@ -0,0 +1,142 @@
+{
+    /// <summary>
+    /// Manager which controls adding pip fragments to the graph.
+    /// </summary>
+    public class PipGraphFragmentManager : IPipGraphFragmentManager
+    {
+        private ConcurrentDictionary<string, (PipGraphFragmentSerializer, Task<bool>)> m_readFragmentTasks = new ConcurrentDictionary<string, (PipGraphFragmentSerializer, Task<bool>)>();
+
+        private IPipGraph m_pipGraph;
+
+        private PipExecutionContext m_context;
+
+        private PipGraphFragmentContext m_fragmentContext;
+
+        private LoggingContext m_loggingContext;
+
+        /// <summary>
+        /// PipGraphFragmentManager
+        /// </summary>
+        public PipGraphFragmentManager(LoggingContext loggingContext, PipExecutionContext context, IPipGraph pipGraph)
+        {
+            m_loggingContext = loggingContext;
+            m_context = context;
+            m_pipGraph = pipGraph;
+            m_fragmentContext = new PipGraphFragmentContext();
+        }
+
+        /// <summary>
+        /// Add a single pip graph fragment to the graph.
+        /// </summary>
+        /// <param name="fragmentName">Name of the fragment, for use with specifying dependencyNames</param>
+        /// <param name="filePath">Path of binary fragment file</param>
+        /// <param name="dependencyNames">Names of fragments it depends on</param>
+        /// <returns>Task that returns true if successfully added all pips in the fragment to the graph</returns>
+        public Task<bool> AddFragmentFileToGraph(string fragmentName, AbsolutePath filePath, string[] dependencyNames)
+        {
+            foreach (var dependencyName in dependencyNames)
+            {
+                Contract.Assert(m_readFragmentTasks.ContainsKey(dependencyName), "Fragment has not yet been added for dependency: " + dependencyName);
+            }
+
+            var deserializer = new PipGraphFragmentSerializer();
+            Task<bool> readFragmentTask = Task.Run(() =>
+            {
+                Task.WaitAll(dependencyNames.Select(dependencyName => m_readFragmentTasks[dependencyName].Item2).ToArray());
+                return deserializer.Deserialize(fragmentName, m_context, m_fragmentContext, filePath, (Pip p) => AddPipToGraph(fragmentName, p));
+            });
+
+            m_readFragmentTasks[fragmentName] = (deserializer, readFragmentTask);
+            return readFragmentTask;
+        }
+
+        /// <summary>
+        /// GetAllFragmentTasks
+        /// </summary>
+        public IReadOnlyCollection<(PipGraphFragmentSerializer, Task<bool>)> GetAllFragmentTasks()
+        {
+            return m_readFragmentTasks.Select(x => x.Value).ToList();
+        }
+
+        private bool AddPipToGraph(string fragmentName, Pip pip)
+        {
+            try
+            {
+                PipId originalPipId = pip.PipId;
+                pip.ResetPipIdForTesting();
+                bool added = false;
+                switch (pip.PipType)

```

Comment:
Do you need to handle `PipType.Ipc` here?

Reply:
good point

---
In reply to: [289542983](https://github.com/microsoft/BuildXL/pull/394#discussion_r289542983) [](ancestors = 289542983)